bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/scheduler2/CronAdjuster.java

```diff
@@ -0,0 +1,551 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.core.internal.scheduler2;
+
+import java.time.DayOfWeek;
+import java.time.temporal.ChronoField;
+import java.time.temporal.ChronoUnit;
+import java.time.temporal.Temporal;
+import java.time.temporal.TemporalAdjuster;
+import java.time.temporal.TemporalField;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * This class creates a Temporal Adjuster that takes a temporal and adjust it to
+ * the next deadline based on a cron specification.
+ *
+ * @See http://www.cronmaker.com/
+ * @See http://www.quartz-scheduler.org/documentation/quartz-1.x/tutorials/
+ *      crontrigger
+ * @author Peter Kriens - initial contribution and API
+ */
+
+class CronAdjuster implements TemporalAdjuster {
+
+    static Pattern WEEKDAY_P = Pattern.compile("(?<day>\\d+|MON|TUE|WED|THU|FRI|SAT|SUN)(#(?<nr>\\d+)|(?<l>L))?",
+            Pattern.CASE_INSENSITIVE);
+
+    /*
+     * A function interface that we use to check a Temporal to see if it matches
+     * a part of the specification. We combine these checkers in and and ors.
+     */
+    interface Checker {
+        boolean matches(Temporal t);
+    }
+
+    Checker ALWAYS_FALSE = (t) -> false;
+    Checker ALWAYS_TRUE = (t) -> true;
+
+    /*
+     * Maintains the type and the combined checker. It can verify if a specific
+     * part of the temporal is ok, and if not, it will reset it to the next
+     * higher temporal with the lower fields set to their minimum value.
+     */
+    static class Field {
+        ChronoField type;
+        Checker checker;
+
+        Temporal isOk(Temporal t) {
+            if (checker.matches(t)) {
+                return null;
+            }
+
+            Temporal out = t.plus(1, type.getBaseUnit());
+
+            switch (type) {
+                case YEAR:
+                    out = out.with(ChronoField.MONTH_OF_YEAR, 1);
```

Comment:
Missing breaks or intentional, if intentional maybe document that, because fall through is not that often used on purpose